### ISSUE TITLE
NixOS 23.11 — Feature Freeze & Release Blockers

### ISSUE THREAD

**Issue by figsoda** [NixOS/nixpkgs#259040]
Hi, we are [figsoda](https://github.com/figsoda) & [Ryan Lahfa](https://github.com/RaitoBezarius), the [release managers](https://nixos.github.io/release-wiki/Release-Managers.html) for NixOS 23.11 ("_Tapir_").

As we approach the next release of NixOS, it's time to discuss the upcoming feature freeze period. We want to ensure that this release is as stable and reliable as possible, and your contributions are crucial to achieving that goal, here are the two upcoming dates in that context:

- **2023-10-16**: Restrict breaking changes to [Release Critical Packages](https://nixos.github.io/release-wiki/Release-Critical-Packages.html)
- **2023-10-30**: Restrict all breaking changes except for desktop environments.

The complete timeline is available here:

- https://github.com/NixOS/nixpkgs/issues/258640

The first restriction will be in place in 12 days, therefore, we encourage all teams to check in now and mention possible roadblocks, so we get a good overview of what's ahead.

# Blockers

Whether you were pinged or not, we encourage everyone to create issues for release blockers and add them to the [23.11 Blockers project](https://github.com/orgs/NixOS/projects/24) (It is not public yet, so you might not be able to see this if you are not in the organization).

# Check-in

Please comment your status quo and possible issues as soon as possible. We'll check teams once they have checked in.

If you think some subsystem, person or team should be added or removed from this list, update `maintainers/team-list.nix` in time before the next feature freeze announcement.

# Desktop environments

- [ ] Cinnamon: @bobby285271 @mkg20001
- [ ] Deepin: @wineee
- [ ] Enlightenment: @NixOS/enlightenment
- [ ] GNOME: @NixOS/gnome @bobby285271 @dasj19 @amaxine
- [ ] LXQt: @NixOS/lxqt
- [ ] Lomiri: @OPNA2608
- [ ] Lumina: @NixOS/lumina
- [ ] MATE: @bobby285271 @johannesloetzsch @romildo
- [ ] Pantheon: @NixOS/pantheon
- [ ] Qt / KDE: @NixOS/qt-kde
- [ ] Xfce: @bobby285271 @romildo @muscaln

# Platforms

- [ ] Darwin: @NixOS/darwin-maintainers

# Languages ecosystem

- [ ] BEAM: @NixOS/beam @minijackson
- [ ] C: @matthewbauer @Mic92
- [ ] Coq: @CohenCyril @Zimmi48 @siraben @vbgl @Alizter
- [ ] Dhall: @Gabriella439 @ehmry
- [ ] Go: @NixOS/golang
- [ ] Haskell: Haskell: @NixOS/haskell
- [ ] Lisp: @NixOS/lisp @hraban
- [ ] Lua: @NixOS/lua
- [ ] Node.js: @lilyinstarlight @marsam @winterqt
- [ ] PHP: @NixOS/php
- [ ] Perl: @stigtsp
- [ ] Python: @FRidh @mweinelt @jonringer @tjni
- [ ] R: @bcdarwin @jbedo
- [ ] Ruby: @marsam
- [ ] Rust: @NixOS/rust
- [ ] Zig: @AndersonTorres @figsoda

# Compilers, build systems

- [ ] Bazel: @mboes @marsam @uri-canva @avdv @olebedev @groodt @aherrmann @ylecornec
- [ ] LLVM: @dtzWill @Ericson2314 @lovek323 @primeos @alyssais @RaitoBezarius @rrbutani @sternenseemann

# Modules

- [ ] ACME: @aanderse @andrew-d @arianvp @emilazy @flokli @m1cr0man
- [ ] systemd: @NixOS/systemd

# Internals

- [ ] Docs: @asymmetric @ryantm
- [ ] Marketing: @garbas @tomberek
- [ ] Module system: @infinisil @roberth
- [ ] Nix/nix-cli ecosystem: @edolstra @grahamc @nbp
- [ ] NixOS tests: @tfc

# Everyone Else

@NixOS/nixpkgs-committers @NixOS/release-engineers

# Finally

No issue is too big or too small, but let's remember that we are all working on the project voluntarily in our free time here, so let's focus on issues that can be realistically addressed in the remaining time before the release.

We thank everyone for their contribution!

**Comment by mweinelt** [comment]
Python 3.11 is our target major/minor for the upcoming release. The upgrade is currently on staging-next and will arrive on master shortly.

I plan on doing another python-updates cycle before the freeze.

We have no blockers right now.

**Comment by figsoda** [comment]
More pings since GitHub limits the amount of pings, please refer to the original issue for the checkboxes

# Modules

- ACME: [...] @emilazy @flokli @m1cr0man

# Internals

- Docs: @asymmetric @ryantm
- Marketing: @garbas @tomberek
- Module system: @infinisil @roberth
- Nix/nix-cli ecosystem: @edolstra @grahamc @nbp
- NixOS tests: @tfc

**Comment by RaitoBezarius** [comment]
LLVM here, we have no blockers to the best of my knowledge, LLVM17 is being developed in #258614, I do not plan to push for LLVM17 for this release, it may or may not land in time. We still have technical debt to pay for LLVM git.

**Comment by stigtsp** [comment]
perlPackages need a major update, I'm working on this.

**Comment by ElvishJerricco** [comment]
systemd-initrd is of course not a blocker at all, but I *would like* to try and finally get it feature complete so we can remove the "experimental" wording from the option description before release. It would allow us to start working towards making it the default in a future release. The biggest obstacle to this is probably the unsupported options in `luksroot.nix`, such as yubikey related things.

**Comment by hraban** [comment]
Lisp's SBCL & CLisp are in a good place. Good timing for CLisp actually which just became available on x86_64-darwin and aarch64-darwin.

I'm not involved with the lispPackages scope so I won't speak to that.

**Comment by etu** [comment]
As for PHP we believe it should be good to go, we recently integrated a new project builder which did have some initial bugs that have all been addressed and I believe that we're in a good place as of now :slightly_smiling_face:

**Comment by NickCao** [comment]
Qt 6.6 is scheduled to release on 2023/10/10, but since Qt 6.5 is the LTS release, we may keep that for 23.11 as well.

**Comment by reckenrode** [comment]
I’d like to see the clang 16 bump for Darwin make it for 23.11. The cross stuff would also be a nice bonus. I don’t know how realistic the stdenv stuff is. I’ve been working on it since May, and the update PR has been open since July, but there’s still breakage to fix.

If it the bump doesn’t make 23.11, it can slip to 24.05, but then the plan would be to bump it again to clang 18 in 24.11 (to put Darwin on an annual update cycle). Going to clang 17 for 24.05 isn’t really an option currently because (as far as I know) libc++ 17 dropped support for the 10.12 SDK and requires 10.13.

**Comment by OPNA2608** [comment]
Not enough Lomiri DE stuff ended up in Nixpkgs to be usable or have any blockers yet.

**Comment by Gabriella439** [comment]
No blockers for Dhall

**Comment by SuperSandro2000** [comment]
#244742 is a blocker and the workarounds from Ubuntu/Debian will not work for us since we can't predict rebuilds of mesa.

**Comment by infinisil** [comment]
No blockers for the module system

**Comment by jbedo** [comment]
No blockers for R.

**Comment by Alizter** [comment]
OCaml looks to be missing from the list. cc @ulrikstrid @vbgl

**Comment by bobby285271** [comment]
We will likely have [Cinnamon](https://github.com/NixOS/nixpkgs/commits/master/pkgs/desktops/cinnamon) (5.6 -> 5.8), [MATE](https://github.com/NixOS/nixpkgs/commits/master/pkgs/desktops/mate) (1.26), [Pantheon](https://github.com/NixOS/nixpkgs/commits/master/pkgs/desktops/pantheon) (7 -> 7.1), [Xfce](https://github.com/NixOS/nixpkgs/commits/master/pkgs/desktops/xfce) (4.18) in this release.

[GNOME 45](https://github.com/NixOS/nixpkgs/pull/247766/files) is a WIP. There is no ETA for the merge.

**Comment by ulrikstrid** [comment]
> OCaml looks to be missing from the list. cc @ulrikstrid @vbgl

I don't have any blockers on my list for OCaml

**Comment by happysalada** [comment]
No blockers for beam. 
We have a [problem with latest macos sonoma](https://github.com/NixOS/nixpkgs/issues/258100) but I don't think it's a blocker (especially since we don't have anyone looking into this, since there isn't enough darwin talent).

**Comment by jonringer** [comment]
I'm opening up my server to any committer for usage on reviewing nixpkgs PRs. [Link to repo](https://github.com/jonringer/server-configuration). Instructions in README, I'll send you IP + port for ssh access.

Specs:
- 128T/64C CPU (3990X)
- 256GB RAM

**Comment by teto** [comment]
Nothing blocking really but I want to get into 23.11 
- a refacto to make neovim configuration  more discoverable (no breaking change, just leveraging structured attributes)
- an update of the lua package set that has been blocked by various broken packages on luarocks.org. I've been working towards fixing those for the past few months and the situation has recently been unblocked so I should be able to bump the set next week.

**Comment by Madouura** [comment]
AFAIK the current ROCm derivations and dependents we have work, so there shouldn't be any blockers there.
It would be nice to get #258328 and maybe #259068 in before breaking changes are restricted however.

**Comment by wineee** [comment]
deepin desktop v23 stable will not be released until 2024

[PR upgrade to v23 beta](https://github.com/NixOS/nixpkgs/pull/257400) is ready for review

**Comment by figsoda** [comment]
> OCaml looks to be missing from the list. cc @ulrikstrid @vbgl

Added. It wasn't included initially because it doesn't have an entry in `team-list.nix`, #260469 should do the trick

**Comment by figsoda** [comment]
No blockers for zig

**Comment by asymmetric** [comment]
@figsoda I think it that for Docs, instead of pinging me, we should ping @NixOS/Documentation-Team

**Comment by minijackson** [comment]
Can ZFS 2.2.0 be part of the 23.11 release? That would be really cool.

**Comment by lilyinstarlight** [comment]
I think the only Node.js blocker would have been #260011 since that was a painful problem for some, but that's been merged now. So no blockers from the Node.js team unless @marsam has something

**Comment by roberth** [comment]
> Node.js

Fetchers suffer from [false successes](https://nixos.org/manual/nixpkgs/stable/#chap-pkgs-fetchers-caveats), so I would highly recommend to write automated tests using [invalidateFetcherByDrvHash](https://nixos.org/manual/nixpkgs/stable/#tester-invalidateFetcherByDrvHash).

**Comment by lilyinstarlight** [comment]
> Fetchers suffer from [false successes](https://nixos.org/manual/nixpkgs/stable/#chap-pkgs-fetchers-caveats), so I would highly recommend to write automated tests using [invalidateFetcherByDrvHash](https://nixos.org/manual/nixpkgs/stable/#tester-invalidateFetcherByDrvHash).

It already has [several tests that use that exact function](https://github.com/NixOS/nixpkgs/blob/493a816d1aa4996bd6a4c72e2b140c558c4bb9ef/pkgs/build-support/node/fetch-npm-deps/default.nix#L26-L114). Do you have specific tests you feel it needs added to these? (Also it might be best to open a new issue to avoid noise on this thread about 23.11)

**Comment by toonn** [comment]
@figsoda, @RaitoBezarius, we merged [the LLVM bump](https://github.com/NixOS/nixpkgs/pull/241692) so it no longer holds up the schedule.

**Comment by figsoda** [comment]
Feature freeze is scheduled to be in exactly 1 week (2023-10-30). If you have any breaking changes you want to see in 23.11, feel free to request reviews and try to get them merged in time!

**Comment by mweinelt** [comment]
We need a fix for the regressions around the postgresql 15 bump and ensureUsers.

https://github.com/NixOS/nixpkgs/issues/216989

**Comment by wineee** [comment]
As For DDE(deepin desktop environment) 

DDE 20 is not maintained by deepin currently, and most components will not get updated

DDE 23 is still in beta release and there will be no stable version in the short term. Beta 3 will be released next year, and it is likely that the stable version will be available in 2025.
Many components are being rewritten or ready to be rewritten, such as a [new window manager](https://github.com/vioken/waylib) based on wlroots, new dock/launcher/control-center based on qml， abandon go package and use c++ to reimplement...

archlinux and UbuntuDDE have ported DDE 23, and I tend not to wait for stable to be released. This will take too long as DDE 20 is already outdated

Actually, I don't want to block 23.11 Release. It would be acceptable to merge on 24.05 (unstable), but I do need to request reviews. The effective feedback received of PR https://github.com/NixOS/nixpkgs/pull/257400  is too few

**Comment by Yarny0** [comment]
>If you have any breaking changes you want to see in 23.11, feel free to request reviews and try to get them merged in time!

It would be nice to have this RFC 42 migration in the next release: https://github.com/NixOS/nixpkgs/pull/253428

**Comment by the-furry-hubofeverything** [comment]
> Feature freeze is scheduled to be in exactly 1 week (2023-10-30). If you have any breaking changes you want to see in 23.11, feel free to request reviews and try to get them merged in time!

It may seem impatient (since I recently got reviews), but It's almost 2 days away from feature freeze and I would really like to see my PR merged (#257780). It's been in the works since the beginning of the month, and maintainers haven't responded for more than two weeks since I asked for reviews.

It introduces multiple blender versions to be included in nixpkgs, since Blender usually release a latest version, and LTS versions that are actively maintained for two years. 

This PR also includes a breaking change, where it removes `blender-with-packages` in favor of `blender.withPackages`, since the inputs are reworked and incompatible with the current usage of `blender-with-packages`. 

This change is important because previously, `blender-with-packages` doesn't provide a way to track the pythonPackages used, something like `blender-with-packages { packages = with blender.python.pkgs; [ bpycv ]; }`. This could be an issue with multiple blender versions, since different versions of blender could use different versions of python, depending on which [VFX reference platform](https://vfxplatform.com/) blender is tracking for that version.

Blender LTS versions lasts two years, so it's highly likely that two different versions of blender uses two different versions of python.

Another drawback of the current implementation is that users might use something like `blender-with-packages { packages = with python3.pythonPackages; [ bpycv ]; }`, which might not match blender's pythonPackage scope (blender 3.3 and 3.6 uses 3.10, future versions that track VFX reference CY2024 would use 3.11). 

My original plan is to point `blender-with-packages` to the latest version of `blender.withPackages`, but with this change, previous configurations wouldn't work with it anyway, so might as well remove it.

This PR reworks `blender-with-packages` to `blender.withPackges` - which can be used similar to how `python.withPackages` is used, and it's more specific to the version of blender used, for example `blenderVersions.blender_3_3.withPackages(ps: [ ps.bpycv ])`

**Comment by Yarny0** [comment]
> Feature freeze is scheduled to be in exactly 1 week (2023-10-30). If you have any breaking changes you want to see in 23.11, feel free to request reviews and try to get them merged in time!

https://github.com/NixOS/nixpkgs/pull/261319 fixes user-controlled lingering which got broken by https://github.com/NixOS/nixpkgs/pull/260248 .

**Comment by figsoda** [comment]
> It may seem impatient (since I recently got reviews), but It's almost 2 days away from feature freeze and I would really like to see my PR merged (#257780).

No worries!

Looking at the changes, looks like the only breaking change is `blender-with-packages -> blender.withPackages`. If you think it might make it easier to review, taking this out into a new PR could be an option, and the rest of the changes can go in after feature freeze.

Great work on the PR, I would love to see it in 23.11!

**Comment by the-furry-hubofeverything** [comment]
> 
> No worries!
> 
> Looking at the changes, looks like the only breaking change is `blender-with-packages -> blender.withPackages`. If you think it might make it easier to review, taking this out into a new PR could be an option, and the rest of the changes can go in after feature freeze.
> 
> Great work on the PR, I would love to see it in 23.11!

https://github.com/NixOS/nixpkgs/pull/264292

**Comment by happysalada** [comment]
should the change of default of nodejs from 18 to 20 be included ? https://github.com/NixOS/nixpkgs/pull/262837
(upstream switched lts to 20 end of october).

**Comment by stigtsp** [comment]
No blockers for Perl :)

**Comment by Madouura** [comment]
We should probably get https://github.com/NixOS/nixpkgs/pull/268746 in.
Strict fix/improvement.

**Comment by SuperSandro2000** [comment]
> should the change of default of nodejs from 18 to 20 be included

I think it is a bit late for changing the default version of a big interpreter this late.

**Comment by vcunat** [comment]
23.11 was branched off several hours ago (for reference), so adding more features seems quite unlikely anymore and focus should be on bufixing/stabilization.

**Comment by infinisil** [comment]
> #244742 is a blocker and the workarounds from Ubuntu/Debian will not work for us since we can't predict rebuilds of mesa.

I want to highlight this. After updating today I just ran into this problem with slack (though there [an update](https://github.com/NixOS/nixpkgs/pull/270440) seems to resolve it) and Elements (this exact issue: https://github.com/NixOS/nixpkgs/issues/261997). This is pretty terrible because it breaks the main application people are using to chat in the Nix community for maybe almost everyone.

**Comment by SuperSandro2000** [comment]
This was fixed for chromium and electron releases. I am not sure which version exactly but programs that vendor electron are maybe still affected. The ones using electron 25+ are in my experience no longer affected.

**Comment by Xyz00777** [comment]
> > #244742 is a blocker and the workarounds from Ubuntu/Debian will not work for us since we can't predict rebuilds of mesa.
> 
> I want to highlight this. After updating today I just ran into this problem with slack (though there [an update](https://github.com/NixOS/nixpkgs/pull/270440) seems to resolve it) and Elements (this exact issue: #261997). This is pretty terrible because it breaks the main application people are using to chat in the Nix community for maybe almost everyone.

it also had broken my Bitwarden and Obsidian, fixed with removing the old GPU Shaders but still it was really confusing...
